public/app/features/alerting/unified/components/receivers/grafanaAppReceivers/onCall/useOnCallIntegration.ts

```diff
@@ -0,0 +1,249 @@
+
+  const [validateIntegrationNameQuery, { isFetching: isValidating }] = useLazyValidateIntegrationNameQuery();
+  const [createIntegrationMutation] = useCreateIntegrationMutation();
+
+  const {
+    data: grafanaOnCallIntegrations = [],
+    isLoading: isLoadingOnCallIntegrations,
+    isError: isIntegrationsQueryError,
+  } = useGrafanaOnCallIntegrationsQuery(undefined, { skip: !isAlertingV2IntegrationEnabled });
+
+  const onCallFormValidators = useMemo(() => {
+    return {
+      integration_name: async (value: string) => {
+        try {
+          await validateIntegrationNameQuery(value).unwrap();
+          return true;
+        } catch (error) {
+          if (isFetchError(error) && error.status === 409) {
+            return 'Integration of this name already exists in OnCall';
+          }
+
+          throw error;
+        }
+      },
+      url: (value: string) => {
+        if (!isAlertingV2IntegrationEnabled) {
+          return true;
+        }
+
+        return grafanaOnCallIntegrations.map((i) => i.integration_url).includes(value)
+          ? true
+          : 'Selection of existing OnCall integration is required';
+      },
+    };
+  }, [grafanaOnCallIntegrations, validateIntegrationNameQuery, isAlertingV2IntegrationEnabled]);
+
+  const extendOnCalReceivers = useCallback(
+    (receiver: Receiver): Receiver => {
+      if (!isAlertingV2IntegrationEnabled) {
+        return receiver;
+      }
+
+      return produce(receiver, (draft) => {
+        draft.grafana_managed_receiver_configs?.forEach((config) => {
+          if (config.type === ReceiverTypes.OnCall) {
+            config.settings[OnCallIntegrationSetting.IntegrationType] = OnCallIntegrationType.ExistingIntegration;
+          }
+        });
+      });
+    },
+    [isAlertingV2IntegrationEnabled]
+  );
+
+  const createOnCallIntegrations = useCallback(
+    async (receiver: Receiver): Promise<Receiver> => {
+      if (!isAlertingV2IntegrationEnabled) {
+        return receiver;
+      }
+
+      const onCallIntegrations = receiver.grafana_managed_receiver_configs?.filter((c) => c.type === 'oncall') ?? [];
+      const newOnCallIntegrations = onCallIntegrations.filter(
+        (c) => c.settings[OnCallIntegrationSetting.IntegrationType] === OnCallIntegrationType.NewIntegration
+      );
+
+      const createNewOnCallIntegrationJobs = newOnCallIntegrations.map(async (c) => {
+        const newIntegration = await createIntegrationMutation({
+          integration: GRAFANA_ONCALL_INTEGRATION_TYPE,
+          verbal_name: c.settings[OnCallIntegrationSetting.IntegrationName],
+        }).unwrap();
+
+        c.settings['url'] = newIntegration.integration_url;
+      });
+
+      await Promise.all(createNewOnCallIntegrationJobs);
+
+      return produce(receiver, (draft) => {
+        draft.grafana_managed_receiver_configs?.forEach((c) => {
+          if (c.type === ReceiverTypes.OnCall) {
+            delete c.settings[OnCallIntegrationSetting.IntegrationType];
```

Comment:
This seems a bit weird – why are we deleting the settings key of both the integration type and the name?

Reply:
The `settings` object is a bag for different properties from the backend perspective. It keeps anything FE sends there.
If we keep FE-only settings there, we might encounter some weird issues - IntegrationType, in this case, is set to `new_oncall_integration` but after saving, it is no longer `new`.
These two properties are a purely FE thing and shouldn't be kept in the backend IMHO